openmc/deplete/integrator/abc.py

```diff
@@ -0,0 +1,268 @@
+
+        self.power = power
+
+    @abstractmethod
+    def __call__(self, conc, rates, dt, power, i):
+        """Perform the integration across one time step
+
+        Parameters
+        ----------
+        conc : numpy.ndarray
+            Initial concentrations for all nuclides in [atom]
+        rates : openmc.deplete.ReactionRates
+            Reaction rates from operator
+        dt : float
+            Time in [s] for the entire depletion interval
+        power : float
+            Power of the system [W]
+        i : int
+            Current depletion step index
+
+        Returns
+        -------
+        proc_time : float
+            Time spent in CRAM routines for all materials
+        conc_list : list of numpy.ndarray
+            Concentrations at each of the intermediate points with
+            the final concentration as the last element
+        op_results : list of openmc.deplete.OperatorResult
+            Eigenvalue and reaction rates from intermediate transport
+            simulations
+        """
+
+    @property
+    @abstractmethod
+    def _N_STAGES(self):
+        """Number of intermediate transport solutions
+
+        Needed to ensure schemes are consistent with restarts
+        """
+
+    def __iter__(self):
+        return zip(self.timesteps, self.power)
+
+    def __len__(self):
+        return len(self.timesteps)
+
+    def _get_bos_data_from_openmc(self, step_index, step_power, bos_conc):
+        x = deepcopy(bos_conc)
+        res = self.operator(x, step_power)
+        self._write_statepoint(step_index)
+        return x, res
+
+    def _get_bos_data_from_restart(self, step_index, step_power, bos_conc):
+        res = self.operator.prev_res[-1]
+        # Depletion methods expect list of arrays
+        # See Operator.__call__, Operator.initial_condition
+        bos_conc = list(res.data[0])
+        rates = res.rates[0]
+        k = ufloat(res.k[0, 0], res.k[0, 1])
+
+        # Scale rates by ratio of powers
+        rates *= step_power / res.power[0]
+        return bos_conc, OperatorResult(k, rates)
+
+    def _get_start_data(self):
+        if self.operator.prev_res is None:
+            return 0.0, 0
+        return (self.operator.prev_res[-1].time[-1],
+                len(self.operator.prev_res) - 1)
+
+    def integrate(self):
+        """Perform the entire depletion process across all steps"""
+        with self.operator as conc:
+            t, self._ires = self._get_start_data()
+
+            for i, (dt, p) in enumerate(self):
+                if i > 0 or self.operator.prev_res is None:
+                    conc, res = self._get_bos_data_from_openmc(i, p, conc)
+                else:
+                    conc, res = self._get_bos_data_from_restart(i, p, conc)
+                proc_time, conc_list, res_list = self(conc, res.rates, dt, p, i)
+
+                # Insert BOS concentration, transport results
+                conc_list.insert(0, conc)
+                res_list.insert(0, res)
```

Comment:
Two possible alternatives
1) Use slice insertion: `conc_list[0:0] = [conc]`
2) Use a [`deque`](https://docs.python.org/3.7/library/collections.html#collections.deque)

Small timeit comparison:
```
$ python -m timeit -n 100 -r 5 -s "l = list(range(100))" "l.insert(0, 0)"
100 loops, best of 5: 160 nsec per loop

$ python -m timeit -n 100 -r 5 -s "l = list(range(100))" "l[0:0] = [0]"
100 loops, best of 5: 142 nsec per loop

$ python -m timeit -n 100 -r 5 -s "from collections import deque; l = deque(range(100))" "l.insert(0, 0)"
100 loops, best of 5: 90.3 nsec per loop
```

Using a `deque` allows allows us to enforce the number of elements set by each integrator using the `_num_stages` attribute. The list should only be as large as `_num_stages` + 1, to hold intermediate step data and beginning of step data. 

`deque`s are also larger in terms of memory. The largest number of elements these will have to hold, at least for the concrete classes here, is five elements. 

Each concrete `__call__` method would have to create these deques and return them, otherwise we have to create them inside the integrate method. When you include the creation of the `deque` and `list` into the timing comparisons, both insert methods take about 1 usec, while the slice indexing takes 0.9 usec. Sicne we know how many stages to anticipate, we could produce one `conc_list` / `conc_deque` on the `Integrators`, and `__call__` just fills the instance list. Saves us from having to return the concentrations and results lists

:man_shrugging:

Reply:
Eh, let's not worry about it for now. I thought about suggesting `deque` too but like you said, then the `__call__` method would need to return it. Compared to the time to perform a MC calculation, this is going to be way in the noise anyway.